Sources/Algorithms/Partition.swift

```diff
@@ -204,3 +204,171 @@ extension Collection {
+  ///
+  /// - Note: This algorithm performs a bit slower than the same algorithm on
+  /// `RandomAccessCollection` since the size of the sequence is unknown, unlike
+  /// `RandomAccessCollection`.
+  @inlinable
+  public func partitioned(
+    _ belongsInSecondCollection: (Element) throws -> Bool
+  ) rethrows -> ([Element], [Element]) {
+    return try _partitioned(belongsInSecondCollection)
+  }
+  
+  @inlinable
+  public func _partitioned(
+    _ belongsInSecondCollection: (Element) throws -> Bool
+  ) rethrows -> ([Element], [Element]) {
+    var lhs = ContiguousArray<Element>()
+    var rhs = ContiguousArray<Element>()
+    
+    for element in self {
+      if try belongsInSecondCollection(element) {
+        rhs.append(element)
+      } else {
+        lhs.append(element)
+      }
+    }
+    
+    return _tupleMap((lhs, rhs), { Array($0) })
+  }
+}
+
+extension Collection {
+  // This is a specialized version of the same algorithm on `Sequence` that
+  // avoids reallocation of arrays since `count` is known ahead of time.
+  @inlinable
+  public func partitioned(
+    _ belongsInSecondCollection: (Element) throws -> Bool
+  ) rethrows -> ([Element], [Element]) {
+    guard !self.isEmpty else {
+      return ([], [])
+    }
+    
+    // Since `RandomAccessCollection`s have known sizes (access to `count` is
+    // constant time, O(1)), we can allocate one array of size `self.count`,
+    // then insert items at the beginning or end of that contiguous block. This
+    // way, we don’t have to do any dynamic array resizing. Since we insert the
+    // right elements on the right side in reverse order, we need to reverse
+    // them back to the original order at the end.
+    
+    let count = self.count
+    
+    // The overhead of this implementation isn’t worth it for collections fewer
+    // than 8 elements. Use the simple `Sequence`-based implementation instead.
+    // This constant was determined using benchmarking. More information:
+    // https://github.com/apple/swift-algorithms/pull/152#issuecomment-887130149
+    if count < 8 {
+      return try _partitioned(belongsInSecondCollection)
+    }
+    
+    // Inside of the `initializer` closure, we set what the actual mid-point is.
+    // We will use this to partitioned the single array into two in constant time.
+    var midPoint: Int = 0
+    
+    let elements = try [Element](
+      unsafeUninitializedCapacity: count,
+      initializingWith: { buffer, initializedCount in
+        var lhs = buffer.baseAddress!
+        var rhs = lhs + buffer.count
+        do {
+          for element in self {
+            if try belongsInSecondCollection(element) {
+              rhs -= 1
+              rhs.initialize(to: element)
+            } else {
+              lhs.initialize(to: element)
+              lhs += 1
+            }
+          }
+          
+          let rhsIndex = rhs - buffer.baseAddress!
+          buffer[rhsIndex...].reverse()
```

Comment:
This reverse operation can be deferred until these elements are stored in a newly allocated array, that might speed things up slightly.

Reply:
When would the reversal happen? If removing line 315 here and instead reversing the `ArraySlice` right before its conversion to an `Array`, it makes it slower.

Before:
```swift
          buffer[rhsIndex...].reverse()
          // …
        } catch {
          // …
        }
      })
    
    let lhs = elements[..<midPoint]
    let rhs = elements[midPoint...]
    return (
      Array(lhs),
      Array(rhs)
    )
```

After:
```swift
    let lhs = elements[..<midPoint]
    let rhs = elements[midPoint...].reversed()
    return (
      Array(lhs),
      Array(rhs)
    )
```

<img width="1280" alt="chart" src="https://user-images.githubusercontent.com/963948/129958283-5e7dd1e6-35fd-4941-87ae-adccaf668804.png">